### DISCUSSION TITLE
How does communication occur between SM and PHM?

### DISCUSSION THREAD

**Discussion by Emanonvvv** [langroodi/Adaptive-AUTOSAR#26] [General]
Hello, I gained valuable insights from the previous question. This time, I have a question about communication between SM and PHM that I'd like to ask you. According to the documentation of PHM, PHM should call the RecoveryAction::RecoveryHandler function to report faults to SM. Whether this function can be called depends on the value of the Offer variable within RecoveryAction. The image indicates that it's SM that calls the offer() function to set this boolean value to 1. So, who should instantiate this class?

I noticed that you've overridden RecoveryAction and placed it inside PHM, with the RecoveryHandler function used to log information. How can this error information be communicated to SM?

**Comment by Emanonvvv** [comment]
<img width="458" alt="屏幕截图 2023-08-31 192250" src="https://github.com/langroodi/Adaptive-AUTOSAR/assets/77663843/7efebd81-abb5-43b7-b3cc-87e0d009a36b">
In AUTOSAR_EXP_SWArchitecture 11.1.2.2

**Comment by langroodi** [comment]
First of all, the `LogRecoveryAction` in this repo as the concrete implementation of the abstract `RecoveryAction` has been really simplified which does not interact with SM. It just logs the failure. What I will explain further is just my own understanding from the standard and please do not take it as the absolute truth:

The owner of the recovery action class is PHM. So, PHM needs to instantiate the class. The reason for that is, if the notifying to SM fails, PHM needs to take another recovery action which may interact with its hardware watchdog. So, basically PHM has a bunch of recovery actions which triggers them in different scenarios. However, PHM is not the only one who can control the offering of a recovery action. As you mentioned, SM can control the offering of certain recovery actions.

The SM execution of a recovery action `Offer` and `StopOffer` methods at PHM side can be done via a Remote Procedure Call (RPC) mechanism. The RPC can be implemented through DDS or SOME/IP Request-Response pattern. The notification to SM via `RecoveryHandler`can be implemented also through DDS or SOME/IP Publish-Subscribe pattern. However, both situations have not been considered in this repository.
